tests/msc3930_test.go

```diff
@@ -0,0 +1,182 @@
+			// There should only be one condition defined for this type
+			match.JSONKeyEqual("conditions.#", 1),
+			// Check the contents of the first (and only) condition
+			match.JSONKeyEqual("conditions.0.kind", "event_match"),
+			match.JSONKeyEqual("conditions.0.key", "type"),
+			match.JSONKeyEqual("conditions.0.pattern", "org.matrix.msc3381.poll.start"),
+		)
+
+		// This push rule creates a sound when a poll is ended in any room.
+		pollEndRule := alice.GetPushRule(t, "global", "underride", pollEndRuleID)
+		must.MatchGJSON(
+			t,
+			pollEndRule,
+			// Check that the appropriate actions are set for this rule
+			match.JSONKeyEqual("actions", []string{"notify"}),
+			match.JSONKeyEqual("default", true),
+			match.JSONKeyEqual("enabled", true),
+			// There should only be one condition defined for this type
+			match.JSONKeyEqual("conditions.#", 1),
+			// Check the contents of the first (and only) condition
+			match.JSONKeyEqual("conditions.0.kind", "event_match"),
+			match.JSONKeyEqual("conditions.0.key", "type"),
+			match.JSONKeyEqual("conditions.0.pattern", "org.matrix.msc3381.poll.end"),
+		)
+
+		// The DM-specific rules for poll start and poll end should come before the rules that
+		// define behaviour for any room. We verify this by ensuring that the DM-specific rules
+		// have a lower index when requesting all push rules.
+		allPushRules := alice.GetAllPushRules(t)
+		globalUnderridePushRules := allPushRules.Get("global").Get("underride").Array()
+
+		pollStartOneToOneRuleIndex := math.MaxInt64
+		pollEndOneToOneRuleIndex := math.MaxInt64
+		for index, rule := range globalUnderridePushRules {
+			// Iterate over the user's global underride rules as a client would.
+			// If we come across a one-to-one room rule ID, we note down its index.
+			// When we come across the rule ID of its generic counterpart, we ensure
+			// its counterpart is at a higher index, and thus would be considered
+			// lower priority.
+			if rule.Get("rule_id").Str == pollStartOneToOneRuleID {
+				pollStartOneToOneRuleIndex = index
+			} else if rule.Get("rule_id").Str == pollEndOneToOneRuleID {
+				pollEndOneToOneRuleIndex = index
+			} else if rule.Get("rule_id").Str == pollStartRuleID {
+				if pollStartOneToOneRuleIndex > index {
+					t.Fatalf(
+						"Expected rule '%s' to come after '%s' in '%s's global underride rules",
+						pollStartRuleID,
+						pollStartOneToOneRuleID,
+						alice.UserID,
+					)
+				}
+			} else if rule.Get("rule_id").Str == pollEndRuleID {
+				if pollEndOneToOneRuleIndex > index {
+					t.Fatalf(
+						"Expected rule '%s' to come after '%s' in '%s's global underride rules",
+						pollEndRuleID,
+						pollEndOneToOneRuleID,
+						alice.UserID,
+					)
+				}
+			}
+		}
+	})
+
+	// TODO: Test whether the homeserver correctly calls POST /_matrix/push/v1/notify on the push gateway
+	// in accordance with the push rules. Blocked by Complement not having a push gateway implementation.
```

Comment:
Heads up: I don't intend to fix this in this PR, since it requires the heavy chunk of work of making Complement support handling [the Push Gateway API](https://spec.matrix.org/v1.5/push-gateway-api/). The known implementations of https://github.com/matrix-org/matrix-spec-proposals/pull/3930 are currently intending to process push rules entirely client-side anyways.